runtime/src/main/java/org/corfudb/recovery/FastSmrMapsLoader.java

```diff
@@ -183,12 +183,14 @@ public void updateStreamTails(long address, ILogData logData) {
         // On checkpoint, we also need to track the stream tail of the checkpoint
         if (isCheckPointEntry(logData)) {
             if (logData.getCheckpointType() == CheckpointEntry.CheckpointEntryType.END) {
-                streamTails.put(logData.getCheckpointedStreamId(),
-                        getStartAddressOfCheckPoint(logData));
+                streamTails.compute(logData.getCheckpointedStreamId(),
+                        (uuid, value) -> (value == null) ? getStartAddressOfCheckPoint(logData)
```

Comment:
Dahlia's example:

1 2 3 4 5 6 7
A A A A A A

Where 5 6 7 have CP-start, CP-Cont and CP-End  and then any non-cp entries after that will overwrite the stream tail with a lower address because the backpoints point to an earlier address.